### ISSUE TITLE
Validation of nullable reference types (see also #1648)

### ISSUE THREAD

**Issue by Rahtgaz** [FluentValidation/FluentValidation#2271]
### Is your feature request related to a problem? Please describe.

3 years have passed since #1648 was closed for internal discussions. Will the ability to validate optional types ever be introduced into FluentValidation, or was it abandoned?

### Describe the solution you'd like

Jeremy comes up with an extension method at the end of that discussion. But the compiler does complain about nullability rules not being respected in the declaration of the `NullableChildValidatorAdaptor class.

I'm unsure if those two warnings at `IPropertyValidator<T, TProperty?>, IAsyncPropertyValidator<T,TProperty?>` could eventually result in runtime errors. But if this is simply the result of the compiler not being smart enough, this extension method could be added into a release.

### Describe alternatives you've considered

Currently I prefer not to manuall add the extension method to my project. I'd rather not add code I can't understand or that forces me to learn more about the internal of FluentValidation than I care to know.

Instead, I manage the problem with a small variant of the original hack:

```
public class TimeRangeValidator : AbstractValidator<TimeRangeRequest?>
{
    public TimeRangeValidator()
    {
        RuleFor(x => x!.Min)
            .LessThanOrEqualTo(x => x!.Nominal)
            .When(x => x is not null);

        RuleFor(x => x!.Max)
            .GreaterThanOrEqualTo(x => x!.Nominal)
            .When(x => x is not null);
    }
}
```

And this can now be used with a nullable object instance (`CycleTime` is a nullable `TimeRangeRequest`):

```
RuleFor(x => x.WorkTask.CycleTime)
    .SetValidator(new TimeRangeValidator())
    .When(x => x.WorkTask.CycleTime is not null);
```

### Additional Context

I'm not particularly happy with the hack either because of the use of null-forgiving operator. It will throw an null reference exception if I forget to check for null at the caller side. 

But the code is understandable to anyone reading it. It would be much better however if none of this would be necessary.

**Comment by JeremySkinner** [comment]
Hi, thanks for the suggestion

> Will the ability to validate optional types ever be introduced into FluentValidation, or was it abandoned?

No, this isn't something we'll be introducing at any point. 

> I'm unsure if those two warnings at IPropertyValidator<T, TProperty?>, IAsyncPropertyValidator<T,TProperty?> could eventually result in runtime errors. 

There have to be null-forgiving operators in place *somewhere* in order to support this. These extension methods essentially just push them out of sight - instead of being in the `RuleFor` definition directly, they're hidden by the extension method. Personally I think it's better to have them as part of the `RuleFor` rather than hiding them away, but the extension method provides an alternative (note I haven't tested this approach with FluentValidation 11.x)

 > this extension method could be added into a release.

This isn't something I'm interested in supporting directly I'm afraid, so won't be included in a release.

> It will throw an null reference exception if I forget to check for null at the caller side.

No, this isn't correct. FluentValidation will never invoke child validators for null references. If the `CycleTime` property is null, then the `TimeRangeValidator` won't be invoked.
